File: book/2e/data/ch04/top-words.py
#!/usr/bin/env python
import re
import sys

from collections import Counter
from urllib.request import urlopen

def top_words(text, n):
    with urlopen("https://raw.githubusercontent.com/stopwords-iso/stopwords-en/master/stopwords-en.txt") as f:
        stopwords = f.read().decode("utf-8").split("\n")

    words = re.findall("[a-z']{2,}", text.lower())
    words = (w for w in words if w not in stopwords)

    for word, count in Counter(words).most_common(n):
        print(f"{count:>7} {word}")


if __name__ == "__main__":
    text = sys.stdin.read()

    try:
        n = int(sys.argv[1])
    except:
        n = 10

    top_words(text, n)

File: book/2e/data/ch04/top-words.R
#!/usr/bin/env Rscript
n <- as.integer(commandArgs(trailingOnly = TRUE))
if (length(n) == 0) n <- 10

f_stopwords <- url("https://raw.githubusercontent.com/stopwords-iso/stopwords-en/master/stopwords-en.txt")
stopwords <- readLines(f_stopwords, warn = FALSE)
close(f_stopwords)

f_text <- file("stdin")
lines <- tolower(readLines(f_text))

words <- unlist(regmatches(lines, gregexpr("[a-z']{2,}", lines)))
words <- words[is.na(match(words, stopwords))]

counts <- sort(table(words), decreasing = TRUE)
cat(sprintf("%7d %s\n", counts[1:n], names(counts[1:n])), sep = "")
close(f_text)
